components/suggest/src/store.rs

```diff
@@ -388,14 +422,24 @@ where
 
             match fields {
                 SuggestRecord::AmpWikipedia => {
-                    self.ingest_attachment(writer, record, |dao, record_id, suggestions| {
-                        dao.insert_amp_wikipedia_suggestions(record_id, suggestions)
-                    })?;
+                    self.ingest_attachment(
+                        &SuggestRecordType::AmpWikipedia.last_ingest_meta_key(),
```

Comment:
Q: we already have an argument `last_ingest_meta_key`, why do we need to generate another meta key for each `ingest_attachment()`? Are there different?

I find it quite hard to track things as now we have to pass this `last_ingest_meta_key` all the way to the bottom, e.g. `ingest_attachment() => ingest_record() => dao.handle_ingested_record()` because we call `self.put_last_ingest_if_newer(last_ingest_key)` in all the high-level APIs of `SugegstDao`. I wonder if it'd be easier to do that at the store level, such as in `Store::ingest_record()` so that `last_ingest_meta_key` won't contaminate stuff in `db.rs`.

Reply:
Alternatively, if updating last_ingest_meta_key at the store level is non-trivial to refactor, we can combine this meta key and `record` into a new struct as the latter is already passed around everywhere.

WDYT?

Reply:
Oh, I think this was an artifact of refactoring the code to return multiple types. Now, it's only a singular type, so we can just use the type passed in params. I'll fix this.

> I wonder if it'd be easier to do that at the store level, such as in Store::ingest_record() so that last_ingest_meta_key won't contaminate stuff in db.rs.

This would be messier and kind of works against the refactoring to move logical pieces of logic to the Dao. 

> Alternatively, if updating last_ingest_meta_key at the store level is non-trivial to refactor, we can combine this meta key and record into a new struct as the latter is already passed around everywhere.

It's only one extra value, so I don't feel like it's a huge deal to pass it as a param. If we find that we have a lot of logical pieces of data that we continually pass with records, then I this a new struct would make sense at that point. Since it's only one piece of data, I'm not sure it's worth creating a new object.

> I find it quite hard to track things as now we have to pass this last_ingest_meta_key all the way to the bottom, e.g. ingest_attachment() => ingest_record() => dao.handle_ingested_record() because we call self.put_last_ingest_if_newer(last_ingest_key) in all the high-level APIs of SugegstDao. I

I think the confusion is because `ingest_attachment` has logic that calls `put_last_ingest_if_newer`, which isn't encapsulated in the high level dao functions. Ideally, we have it's just `ingest_record` -> `dao.handle_ingested_record`, which isn't as many levels of indirection. I don't really understand how attachments work, so not sure if there's a way we can refactor this to make it simpler...

I would like to get this PR in as soon as possible and let someone more skilled with Rust refactor it. What do you think about leaving the code as is (with the fix mentioned above), then letting someone else do the refactor?

Reply:
Yup, for sure. Feel free to leave it as is. I think some future refactors (e.g. merge `record` and `last_ingest_meta_key` into an ingestion context struct and pass it down) would make this easier to follow and maintain.

Reply:
Arg, the passing of `last_ingest_key` won't work with the current tests (because the full snapshots are being passed). So, I'll need to keep the full reference for the tests to pass, or otherwise figure out how to fix the tests referenced here: https://github.com/mozilla/application-services/pull/6095#discussion_r1484415054 Thoughts on whether this is a dealbreaker? @ncloudioj I can try to fix the tests instead...

Reply:
Hmm, I am still not entirely sure about the problem, but I'd try fixing the tests to make the code easier to follow here.

Reply:
@mtbottle can you add a brief comment to note why it needs to generate the last ingest meta key again here so that we know how to refactor it in the future. Adding one to here would be sufficient, not need to do so for other match variants. Thanks!